src/liblelantus/lelantus_prover.cpp

```diff
@@ -101,32 +103,78 @@ void LelantusProver::generate_sigma_proofs(
-        Pk[i].resize(params->get_sigma_m());
-        Yk[i].resize(params->get_sigma_m());
-        a[i].resize(params->get_sigma_n() * params->get_sigma_m());
-        sigmaProver.sigma_commit(C_, indexes[i], rA[i], rB[i], rC[i], rD[i], a[i], Tk[i], Pk[i], Yk[i], sigma[i], sigma_proofs[i]);
+
+    std::size_t threadsMaxCount = std::min((unsigned int)N, std::thread::hardware_concurrency());
+    std::vector<std::shared_future<void>> parallelTasks;
+    parallelTasks.reserve(threadsMaxCount);
+    ParallelOpThreadPool<void> threadPool(threadsMaxCount);
+
+    std::vector<std::vector<GroupElement>> C_;
+    C_.resize(N);
+    for (std::size_t j = 0; j < N; j += threadsMaxCount) {
+        for (std::size_t i = j; i < j + threadsMaxCount; ++i) {
+            if (i < N) {
+                if (!c.count(Cin[i].second))
+                    throw std::invalid_argument("No such anonymity set or id is not correct");
+
+                GroupElement gs = (params->get_g() * Cin[i].first.getSerialNumber().negate());
+                serialNumbers.emplace_back(Cin[i].first.getSerialNumber());
+
+                C_[i].reserve(c.size());
+
+                const auto& set = c.find(Cin[i].second);
+                if (set == c.end())
+                    throw std::invalid_argument("No such anonymity set");
+
+                for (auto const &coin : set->second)
+                    C_[i].emplace_back(coin.getValue() + gs);
+
+                rA[i].randomize();
+                rB[i].randomize();
+                rC[i].randomize();
+                rD[i].randomize();
+                Tk[i].resize(params->get_sigma_m());
+                Pk[i].resize(params->get_sigma_m());
+                Yk[i].resize(params->get_sigma_m());
+                a[i].resize(params->get_sigma_n() * params->get_sigma_m());
+
+                auto& sigma_i = sigma[i];
+                auto& rA_i = rA[i];
+                auto& rB_i = rB[i];
+                auto& rC_i = rC[i];
+                auto& rD_i = rD[i];
+                auto& a_i = a[i];
+                auto& Tk_i = Tk[i];
+                auto& Pk_i = Pk[i];
+                auto& Yk_i = Yk[i];
+                auto& prover = sigmaProver;
+                auto& commits = C_[i];
+                auto& index = indexes[i];
+                auto& proof = sigma_proofs[i];
+                parallelTasks.emplace_back(threadPool.PostTask([&]() {
+                    return prover.sigma_commit(commits, index, rA_i, rB_i, rC_i, rD_i, a_i, Tk_i, Pk_i, Yk_i, sigma_i, proof);
+                }));
+            } else
+                break;
+        }
+
+        bool isFail = false;
+        for (auto& th : parallelTasks) {
+            try {
+                th.get();
+            } catch (std::exception& except) {
+                LogPrintf(except.what());
+                isFail = true;
+            }  catch (...) {
+                LogPrintf("Lelantus proof creation failed.");
+                isFail = true;
```

Comment:
look for docs there
https://www.boost.org/doc/libs/1_54_0/doc/html/thread/thread_management.html#thread.thread_management.tutorial.interruption

you can't use `catch(...)` here (without rethrowing at least). This this catch `boost::thread_interrupted` and then throw `std::runtime_error`. Meaning every time you gracefully shutdown there will be runtime error

please use the method from zerocoin code to block thread interruption while in function